File: jempol_turbo/scoring.py
"""Typing metrics and ranking helpers."""

from __future__ import annotations

from dataclasses import dataclass


@dataclass(frozen=True)
class Metrics:
    typed_chars: int
    correct_chars: int
    progress: float
    accuracy: float
    wpm: float
    score: int
    finished: bool


def count_correct_chars(target_text: str, typed_text: str) -> int:
    correct = 0
    for expected, actual in zip(target_text, typed_text):
        if expected == actual:
            correct += 1
    return correct


def compute_metrics(
    target_text: str,
    typed_text: str,
    elapsed_seconds: float,
    *,
    completion_bonus: int = 25,
) -> Metrics:
    typed_text = typed_text[: len(target_text)]
    typed_chars = len(typed_text)
    correct_chars = count_correct_chars(target_text, typed_text)
    total_chars = max(len(target_text), 1)
    elapsed_minutes = max(elapsed_seconds, 0.1) / 60.0

    progress = correct_chars / total_chars
    accuracy = correct_chars / max(typed_chars, 1) * 100.0
    wpm = (correct_chars / 5.0) / elapsed_minutes
    finished = typed_text == target_text
    score = round(wpm * (accuracy / 100.0))
    if finished:
        score += completion_bonus

    return Metrics(
        typed_chars=typed_chars,
        correct_chars=correct_chars,
        progress=round(progress, 4),
        accuracy=round(accuracy, 2),
        wpm=round(wpm, 2),
        score=int(score),
        finished=finished,
    )


def ranking_key(player_state: dict) -> tuple[int, float, float, float, int]:
    """Sort key for player ranking.

    Finished players win over unfinished players. Among finished players,
    lower finish time is better. Remaining ties use WPM, accuracy, and score.
    """

    finished = 1 if player_state.get("finished") else 0
    finish_time = float(player_state.get("finish_time") or 999999.0)
    return (
        finished,
        -finish_time,
        float(player_state.get("wpm", 0.0)),
        float(player_state.get("accuracy", 0.0)),
        int(player_state.get("score", 0)),
    )

File: src/lib/scoring.ts
export type Metrics = {
  typed_chars: number;
  correct_chars: number;
  progress: number;
  accuracy: number;
  wpm: number;
  score: number;
  finished: boolean;
};

export function countCorrectChars(targetText: string, typedText: string): number {
  let correct = 0;
  const limit = Math.min(targetText.length, typedText.length);
  for (let index = 0; index < limit; index += 1) {
    if (targetText[index] === typedText[index]) {
      correct += 1;
    }
  }
  return correct;
}

export function computeMetrics(
  targetText: string,
  typedText: string,
  elapsedSeconds: number,
  completionBonus = 25,
): Metrics {
  const cappedText = typedText.slice(0, targetText.length);
  const typedChars = cappedText.length;
  const correctChars = countCorrectChars(targetText, cappedText);
  const totalChars = Math.max(targetText.length, 1);
  const elapsedMinutes = Math.max(elapsedSeconds, 0.1) / 60;
  const progress = correctChars / totalChars;
  const accuracy = (correctChars / Math.max(typedChars, 1)) * 100;
  const wpm = correctChars / 5 / elapsedMinutes;
  const finished = cappedText === targetText;
  const score = Math.round(wpm * (accuracy / 100)) + (finished ? completionBonus : 0);

  return {
    typed_chars: typedChars,
    correct_chars: correctChars,
    progress: Number(progress.toFixed(4)),
    accuracy: Number(accuracy.toFixed(2)),
    wpm: Number(wpm.toFixed(2)),
    score,
    finished,
  };
}

export function rankPlayers<T extends { finished: boolean; finish_time: number | null; wpm: number; accuracy: number; score: number }>(
  players: T[],
): T[] {
  return [...players].sort((a, b) => {
    if (a.finished !== b.finished) return a.finished ? -1 : 1;
    if (a.finished && b.finished && a.finish_time !== b.finish_time) {
      return (a.finish_time ?? 999999) - (b.finish_time ?? 999999);
    }
    if (a.wpm !== b.wpm) return b.wpm - a.wpm;
    if (a.accuracy !== b.accuracy) return b.accuracy - a.accuracy;
    return b.score - a.score;
  });
}
